### PR TITLE
fix: Prometheus URL label

### LINKED ISSUE #2502: Excessive Prometheus label cardinality issue
**Describe the bug**
Excessive Prometheus metrics cardinality causes server degradation. `MetricsManager` middleware creates a new label for all unique URLs, which results in server resources depletion and eventually, panics (see https://github.com/ory/hydra/blob/master/metrics/prometheus/middleware.go#L30). So even a very short-lived pentest scan can generate hundreds of thousands of new metrics which could cause Hydra, Prometheus, and all related monitoring tools to fail. 

A few examples:
```
hydra_response_time_seconds_count{_shard_id="003",az="useast1a",cluster="test.k8s.local",endpoint="/oauth2/'"/logout",env="prod",instance="10.10.10.10:3333",job="test-auth",k8s_namespace="test-auth",k8s_pod="test-auth-649f5965b8-brqgc",node="ip-10-10-10-10",region="useast1"}
hydra_response_time_seconds_count{_shard_id="003",az="useast1a",cluster="test.k8s.local",endpoint="/oauth2/(nslookup hitflvcefreks16da7.bxss.me||perl -e "gethostbyname('hitflvcefreks16da7.bxss.me')")/logout",env="prod",instance="10.10.10.10:3333",job="test-auth",k8s_namespace="test-auth",k8s_pod="test-auth-649f5965b8-brqgc",node="ip-10-10-10-10",region="useast1"}
hydra_response_time_seconds_count{_shard_id="003",az="useast1a",cluster="test.k8s.local",endpoint="/oauth2/-1 OR 3+427-427-1=0+0+0+1 -- /logout",env="prod",instance="10.10.10.10:3333",job="test-auth",k8s_namespace="test-auth",k8s_pod="test-auth-649f5965b8-brqgc",node="ip-10-10-10-10",region="useast1"}
hydra_response_time_seconds_count{_shard_id="003",az="useast1a",cluster="test.k8s.local",endpoint="/oauth2//%2e%2e%5c%2e%2e%5c%2e%2e%5c%2e%2e%5c%2e%2e%5c%2e%2e%5c%2e%2e%5c%2e%2e%5cetc/passwd",env="prod",instance="10.10.10.10:3333",job="test-auth",k8s_namespace="test-auth",k8s_pod="test-auth-649f5965b8-brqgc",node="ip-10-10-10-10",region="useast1"}
hydra_response_time_seconds_count{_shard_id="003",az="useast1a",cluster="test.k8s.local",endpoint="/oauth2//..%c0%af..%c0%af..%c0%af..%c0%af..%c0%af..%c0%af..%c0%af..%c0%af/etc/passwd",env="prod",instance="10.10.10.10:3333",job="test-auth",k8s_namespace="test-auth",k8s_pod="test-auth-649f5965b8-brqgc",node="ip-10-10-10-10",region="useast1"}
hydra_response_time_seconds_count{_shard_id="003",az="useast1a",cluster="test.k8s.local",endpoint="/oauth2//.../.../.../.../.../.../.../.../etc/passwd",env="prod",instance="10.10.10.10:3333",job="test-auth",k8s_namespace="test-auth",k8s_pod="test-auth-649f5965b8-brqgc",node="ip-10-10-10-10",region="useast1"}
hydra_response_time_seconds_count{_shard_id="003",az="useast1a",cluster="test.k8s.local",endpoint="/oauth2//..\..\..\..\..\..\..\..\etc/passwd",env="prod",instance="10.10.10.10:3333",job="test-auth",k8s_namespace="test-auth",k8s_pod="test-auth-649f5965b8-brqgc",node="ip-10-10-10-10",region="useast1"}
hydra_response_time_seconds_count{_shard_id="003",az="useast1a",cluster="test.k8s.local",endpoint="/oauth2//./WEB-INF/web.xml",env="prod",instance="10.10.10.10:3333",job="test-auth",k8s_namespace="test-auth",k8s_pod="test-auth-649f5965b8-brqgc",node="ip-10-10-10-10",region="useast1"}

```

**Reproducing the bug**
1. Run Hydra app on local
2. Send a request `curl http://localhost:4444/randomRequest`
2. Send a request `curl http://localhost:4445/metrics/prometheus`
3. Response for step 3 has the following metrics:
```
hydra_response_time_seconds_bucket{buildTime="",endpoint="/randomRequest",hash="",version="",le="0.005"} 1
hydra_response_time_seconds_bucket{buildTime="",endpoint="/randomRequest",hash="",version="",le="0.01"} 1
hydra_response_time_seconds_bucket{buildTime="",endpoint="/randomRequest",hash="",version="",le="0.025"} 1
hydra_response_time_seconds_bucket{buildTime="",endpoint="/randomRequest",hash="",version="",le="0.05"} 1
hydra_response_time_seconds_bucket{buildTime="",endpoint="/randomRequest",hash="",version="",le="0.1"} 1
hydra_response_time_seconds_bucket{buildTime="",endpoint="/randomRequest",hash="",version="",le="0.25"} 1
hydra_response_time_seconds_bucket{buildTime="",endpoint="/randomRequest",hash="",version="",le="0.5"} 1
hydra_response_time_seconds_bucket{buildTime="",endpoint="/randomRequest",hash="",version="",le="1"} 1
hydra_response_time_seconds_bucket{buildTime="",endpoint="/randomRequest",hash="",version="",le="2.5"} 1
hydra_response_time_seconds_bucket{buildTime="",endpoint="/randomRequest",hash="",version="",le="5"} 1
hydra_response_time_seconds_bucket{buildTime="",endpoint="/randomRequest",hash="",version="",le="10"} 1
hydra_response_time_seconds_bucket{buildTime="",endpoint="/randomRequest",hash="",version="",le="+Inf"} 1
hydra_response_time_seconds_sum{buildTime="",endpoint="/randomRequest",hash="",version=""} 9.4364e-05
hydra_response_time_seconds_count{buildTime="",endpoint="/randomRequest",hash="",version=""} 1
```

**Expected behavior**
Ideally, only endpoints registered in HTTP routers should show up as a separate label in Prometheus metrics. All other requests should generate metrics with the URL label `unmatched`.

PR with a fix will be submitted soon.

### PR DESCRIPTION
**Pull request by seremenko-wish**
## Related issue
Fixes #2502 

## Proposed changes
Because an existing HTTP router library does not register in Context object any valuable information that can be used for determining whether URL was matched or not, routers need to be registered in `MetricsManager` directly and all requests need to be matched a second time. Once the request URL is matched to any of the endpoints in any of the registered routers, middleware emits a metric with a value equal to the path, otherwise label `{unmatched}` is used. All matched parameters in the path are getting replaced by `{param}` in the metric label, so:
- request to `/clients/12345` and `/clients/randomString` with emit an event with label `/clients/{param}`

Known issues:
- because the same MetricsManager is used for both Public and Admin HTTP servers, and our router does not populate context with route matching information (like port, matched endpoint), requests sent to public API can be matched with admin API endpoint. No solution is available for this problem.
- I am trying to reconstruct the endpoint from the request URL and matched parameters, so the request `/clients/clients` can be matched with endpoint `/clients/:id`, and reconstruction logic would replace all strings `clients` to `{param}`. 
- Ideally, we should analyze the status code, but middleware does not have access to this info (or I can't find a way to get it in the middleware)

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](docs/docs).

### CODE REVIEW

**Review by aeneasr — APPROVED** [review]
This is great! I think we can merge it as is. My only question would be - is it normal to remove these params? I guess you would use tracing to find actual slow queries, not prometheus?

**Comment by aeneasr** [comment]
Do you want this to be merged already or do you want to address the known issues first?

---

Not to self (release log):

```
Currently, prometheus labels are URL-specific leading to high cardinality which can cause OOM and other issues in Ory Hydra, Prometheus, and other systems. This patch addresses that issue by removing cardinality:

- TBD
```

**Comment by seremenko-wish** [comment]
> My only question would be - is it normal to remove these params? I guess you would use tracing to find actual slow queries, not prometheus?

Right, Prometheus is a monitoring tool, and it should not collect data for all unique requested URLs, so my solution is acceptable. 

> Do you want this to be merged already or do you want to address the known issues first?

Until we start to use a more powerful router, known issues can't be fixed. So let's proceed and merge it.

Thank you.

**Comment by aeneasr** [comment]
> Ideally, we should analyze the status code, but middleware does not have access to this info (or I can't find a way to get it in the middleware)

I believe that `rw` is of type `negroni.ResponseWriter` so we have access to status, size and so on:

```go
	var rw http.ResponseWriter
	res := rw.(negroni.ResponseWriter)
	status := res.Status()
	size := res.Size()
```

I also saw your PR in Ory Kratos - thank you! I would say we put the contents of https://github.com/ory/kratos/tree/4b4032681dc177bb2811684c5858844a69f2bb1d/metrics/prometheus into `github.com/ory/x/prometheusx` so that we can just import it in e.g. Ory Kratos and that future patches and improvements are rolled out to all projects!

What do you think?

**Comment by seremenko-wish** [comment]
I will experiment with status codes next week, and see how I can improve existing metrics. It would be nice to add counters per status code per endpoint.

**Comment by seremenko-wish** [comment]
Submitted the PR that refactors the code.

For status codes, I will create a separate ticket/feature request.

File: metrics/prometheus/doc.go [hunk 1]
```diff
@@ -1,26 +0,0 @@
-package prometheus
```
Comment (seremenko-wish) [inline metrics/prometheus/doc.go:1]: @aeneasr what should I do with this file?

**Review by aeneasr — APPROVED** [review]

**Comment by aeneasr** [comment]
Awesome, thank you! :)

**Pull request merged**